pkg/controller/master/upgrade/upgrade_repo.go

```diff
@@ -454,3 +482,63 @@ func (r *Repo) getInfo() (*RepoInfo, error) {
 	}
 	return &info, nil
 }
+
+func (r *Repo) getImageList(version string, imageList map[string]bool) error {
+	imageListURL := fmt.Sprintf("http://%s.%s/harvester-iso/bundle/harvester/images-lists-archive/%s/image_list_all.txt",
+		r.getRepoServiceName(),
+		upgradeNamespace,
+		version,
+	)
+
+	req, err := http.NewRequestWithContext(r.ctx, http.MethodGet, imageListURL, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected status: %s", resp.Status)
+	}
+
+	scanner := bufio.NewScanner(resp.Body)
+	for scanner.Scan() {
+		imageName := scanner.Text()
+		imageList[imageName] = true
+	}
+
+	return scanner.Err()
+}
+
+func (r *Repo) getImagesDiffList() ([]string, error) {
+	previousImageList := make(map[string]bool)
+	currentImageList := make(map[string]bool)
+
+	backoff := wait.Backoff{
+		Steps:    30,
+		Duration: 10 * time.Second,
+		Factor:   1.0,
+		Jitter:   0.1,
+	}
+
+	if err := retry.OnError(backoff, util.IsRetriableNetworkError, func() error {
+		logrus.Infof("Trying to get %s image list", r.upgrade.Status.PreviousVersion)
+		err := r.getImageList(r.upgrade.Status.PreviousVersion, previousImageList)
+		if err != nil {
+			return err
+		}
+		logrus.Infof("Trying to get %s image list", currentVersion)
+		return r.getImageList(currentVersion, currentImageList)
+	}); err != nil {
+		return nil, err
+	}
+
+	diffList := difference(previousImageList, currentImageList)
+	logrus.Debugf("Diff: %v", diffList)
```

Comment:
nit: Let's make it to info to record the deleted images. Plans and jobs are gone after the upgrade.

Reply:
Updated, thanks!